### PR TITLE
Allow players to set the flowers the beekeeper feeds to his bees

### PR DESCRIPTION
**Pull request by ToMe25**
Closes #
Closes #
Closes #

# Changes proposed in this pull request:
- Add an ItemList module to the beekeeper to allow the player to select which flowers to feed to the bees.
- The beekeeper will yell if no flower is selected.
- Currently all flowers are disabled by default, as I was told to make them "opt-in".

Review please

### CODE REVIEW

File: src/main/java/com/minecolonies/apiimp/initializer/ModBuildingsInitializer.java [hunk 1]
```diff
@@ -437,6 +441,13 @@ public static void init(final RegistryEvent.Register<BuildingEntry> event)
                                    .addBuildingModuleProducer(() -> new SettingsModule()
                                                                       .with(AbstractBuildingWorker.BREEDING, new BoolSetting(true))
                                                                       .with(BuildingBeekeeper.MODE, new StringSetting(BuildingBeekeeper.HONEYCOMB, BuildingBeekeeper.HONEY, BuildingBeekeeper.BOTH)), SettingsModuleView::new)
+                                   .addBuildingModuleProducer(() -> new ItemListModule(BUILDING_FLOWER_LIST),  () -> new ItemListModuleView(BUILDING_FLOWER_LIST, COM_MINECOLONIES_COREMOD_REQUEST_FLOWERS, false,
+                                     (buildingView) ->
+                                   {
+                                       Set<ItemStorage> flowers = new HashSet<>();
+                                       ItemTags.FLOWERS.getAllElements().forEach((item) -> flowers.add(new ItemStorage(new ItemStack(item))));
```
Comment (Raycoms) [inline src/main/java/com/minecolonies/apiimp/initializer/ModBuildingsInitializer.java]: maybe hide this in a utility method somewhere.

Comment (ToMe25) [reply src/main/java/com/minecolonies/apiimp/initializer/ModBuildingsInitializer.java]: The cook doesn't either, and the method has a similar length so I assumed this ok.
To which utility class should I move it then, or just a method in this class?
Also should I move the cook one too then?

Comment (Raycoms) [reply src/main/java/com/minecolonies/apiimp/initializer/ModBuildingsInitializer.java]: Could be a static method in the CompatibilityManager, if the cook one is similar then yes, please

Comment (ToMe25) [reply src/main/java/com/minecolonies/apiimp/initializer/ModBuildingsInitializer.java]: OK, I'll move those two then.

Comment (ToMe25) [reply src/main/java/com/minecolonies/apiimp/initializer/ModBuildingsInitializer.java]: Done.

File: src/api/java/com/minecolonies/api/compatibility/CompatibilityManager.java [hunk 2]
```diff
@@ -732,4 +735,34 @@ private void discoverModCompat()
...
+    }
+
+    /**
+     * Gets a function returning all the possible flowers for a beekeeper building view.
+     * 
+     * @return the function.
+     */
+    public static Function<IBuildingView, Set<ItemStorage>> getAllFlowersFunction()
+    {
+        return (buildingView) ->
```
Comment (Raycoms) [inline src/api/java/com/minecolonies/api/compatibility/CompatibilityManager.java]: Ohh, I thought you'd make it (buildingView) -> CompatibilityManager.getAllFlowers()

Comment (ToMe25) [reply src/api/java/com/minecolonies/api/compatibility/CompatibilityManager.java]: I can change it if you prefer it that way, but I thought if i make a new method for this anyways I might as well make the initializer line as short as possible.

Comment (Raycoms) [reply src/api/java/com/minecolonies/api/compatibility/CompatibilityManager.java]: Yeah, I want it clear that is expects a function lambda, but make the lambda short

Comment (ToMe25) [reply src/api/java/com/minecolonies/api/compatibility/CompatibilityManager.java]: Ok, i'll Change that later.

Comment (ToMe25) [reply src/api/java/com/minecolonies/api/compatibility/CompatibilityManager.java]: I haven't done much testing after this change, but I don't see any way for me to mess this up in a way that makes basic functionality work, but kills something else.
So this should be done :)

**Review by Raycoms — DISMISSED** [review]

File: src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBeekeeper.java [hunk 3]
```diff
@@ -227,6 +234,19 @@ public int getMaximumHives()
...
 
+    /**
+     * Get a list of all the enabled flowers to use for breeding bees.
+     * 
+     * @return the flowers to use for the beekeeper.
+     */
+    public Set<Item> getAllowedFlowers()
+    {
+        return getModuleMatching(ItemListModule.class, m -> m.getId().equals(BUILDING_FLOWER_LIST)).getList().stream()
+                     .map(ItemStorage::getItemStack)
```
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBeekeeper.java]: Not 100% sure, but couldn't we save a lot of cycles by just having 1 map that does it to item directly?

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/beekeeper/EntityAIWorkBeekeeper.java [hunk 4]
```diff
@@ -148,11 +151,21 @@ private IAIState prepareForHerding()
...
         }
+
         if (!getOwnBuilding().getHarvestTypes().equals(BuildingBeekeeper.HONEYCOMB))
         {
             checkIfRequestForItemExistOrCreateAsynch(new ItemStack(Items.GLASS_BOTTLE));
         }
-        checkIfRequestForTagExistOrCreateAsynch(ItemTags.FLOWERS, 16);
+
+        Set<Item> allowedFlowers = getOwnBuilding().getAllowedFlowers();
+        if (!InventoryUtils.hasItemInItemHandler(worker.getInventoryCitizen(), (stack) -> allowedFlowers.contains(stack.getItem()))
```
Comment (someaddons) [inline src/main/java/com/minecolonies/coremod/entity/ai/citizen/beekeeper/EntityAIWorkBeekeeper.java]: if you keep the list of itemstorages you can search much easier for them,  as our inventories do contain itemstorages too

File: src/api/java/com/minecolonies/api/compatibility/CompatibilityManager.java [hunk 5]
```diff
@@ -732,4 +733,28 @@ private void discoverModCompat()
...
+
+    /**
+     * Gets all the possible foods for the cook.
+     * 
+     * @return a set of the foods.
+     */
+    public static Set<ItemStorage> getAllFoods()
+    {
+         Set<ItemStorage> edibles = IColonyManager.getInstance().getCompatibilityManager().getEdibles();
+         edibles.removeIf(item -> ISCOOKABLE.test(item.getItemStack()));
```
Comment (Raycoms) [inline src/api/java/com/minecolonies/api/compatibility/CompatibilityManager.java]: I double checked in the code, this is all not necessary, IColonyManager.getInstance().getCompatibilityManager().getEdibles() is enough, no need for the filter.

File: src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBeekeeper.java (hunk 3 above)
Comment (ToMe25) [reply src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBeekeeper.java]: I would have probably used getItem, if I didn't overlook that method when making this when implementing that.
I will update that soon.

File: src/api/java/com/minecolonies/api/compatibility/CompatibilityManager.java (hunk 5 above)
Comment (ToMe25) [reply src/api/java/com/minecolonies/api/compatibility/CompatibilityManager.java]: OK, I thought "It was there before, I'm sure theres a reason", but I'll remove it then.

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/beekeeper/EntityAIWorkBeekeeper.java (hunk 4 above)
Comment (ToMe25) [reply src/main/java/com/minecolonies/coremod/entity/ai/citizen/beekeeper/EntityAIWorkBeekeeper.java]: how would I do that?
I tried looking for how that can be done, but I couldn't find it

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/beekeeper/EntityAIWorkBeekeeper.java [hunk 6]
```diff
@@ -242,7 +262,7 @@ private IAIState decideWhatToDo()
 
         final boolean hasBreedingItem =
           InventoryUtils.hasItemInItemHandler((worker.getInventoryCitizen()),
-            (ItemStack stack) -> stack.getItem().isIn(ItemTags.FLOWERS));
+            (ItemStack stack) -> getOwnBuilding().getAllowedFlowers().contains(stack.getItem()));
```
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/entity/ai/citizen/beekeeper/EntityAIWorkBeekeeper.java]: same "isItemInList"

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/beekeeper/EntityAIWorkBeekeeper.java [hunk 7]
```diff
@@ -494,7 +514,7 @@ public boolean equipBreedItem(final Hand hand)
         if (checkIfRequestForTagExistOrCreateAsynch(ItemTags.FLOWERS, 2))
         {
             worker.getCitizenItemHandler()
-              .setHeldItem(hand, InventoryUtils.findFirstSlotInItemHandlerWith(getInventory(), stack -> stack.getItem().isIn(ItemTags.FLOWERS)));
+              .setHeldItem(hand, InventoryUtils.findFirstSlotInItemHandlerWith(getInventory(), stack -> getOwnBuilding().getAllowedFlowers().contains(stack.getItem())));
```
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/entity/ai/citizen/beekeeper/EntityAIWorkBeekeeper.java]: same "isITemInList"

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/beekeeper/EntityAIWorkBeekeeper.java (hunk 4 above)
Comment (Raycoms) [reply src/main/java/com/minecolonies/coremod/entity/ai/citizen/beekeeper/EntityAIWorkBeekeeper.java]: The predicate here should ratherbe "itemlistmodule.isItemInList", there is no need to collect the flowerlist here.

Then you can scrap the method in the building actually and down here just module.getList().mapStorageToStack

Comment (someaddons) [reply src/main/java/com/minecolonies/coremod/entity/ai/citizen/beekeeper/EntityAIWorkBeekeeper.java]: dont we have some itemstorage specific functions which avoids us needing to make it stacks in the first place?

Comment (someaddons) [reply src/main/java/com/minecolonies/coremod/entity/ai/citizen/beekeeper/EntityAIWorkBeekeeper.java]: seems we're missing some itemstorage predicate util for it

Comment (Raycoms) [reply src/main/java/com/minecolonies/coremod/entity/ai/citizen/beekeeper/EntityAIWorkBeekeeper.java]: Not for the citizen inventory, but there is a isInBuilding check afaik for the building.

Comment (ToMe25) [reply src/main/java/com/minecolonies/coremod/entity/ai/citizen/beekeeper/EntityAIWorkBeekeeper.java]: I will probably update this later today to `module.isItemInList` then.

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/beekeeper/EntityAIWorkBeekeeper.java (hunk 7 above)
Comment (ToMe25) [reply src/main/java/com/minecolonies/coremod/entity/ai/citizen/beekeeper/EntityAIWorkBeekeeper.java]: Done.

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/beekeeper/EntityAIWorkBeekeeper.java (hunk 6 above)
Comment (ToMe25) [reply src/main/java/com/minecolonies/coremod/entity/ai/citizen/beekeeper/EntityAIWorkBeekeeper.java]: Done.

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/beekeeper/EntityAIWorkBeekeeper.java (hunk 4 above)
Comment (ToMe25) [reply src/main/java/com/minecolonies/coremod/entity/ai/citizen/beekeeper/EntityAIWorkBeekeeper.java]: Like this?

**Review by someaddons — DISMISSED** [review]

**Review by Raycoms — DISMISSED** [review]

**Pull request merged**